vnext/Desktop/Modules/WebSocketModule.cpp

```diff
@@ -13,123 +14,173 @@
+            options.emplace(Utf8ToUtf16(header.first.getString()), header.second.getString());
+          }
+        }
+
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 3), jsArgAsString(args, 0));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Connect(protocols, options);
+        }
+      }),
+    Method(
+      "close",
+      [this](dynamic args) // [int64_t code, string reason,] int64_t id
+      {
+        // See react-native\Libraries\WebSocket\WebSocket.js:_close
+        if (args.size() == 3) // WebSocketModule.close(statusCode, closeReason, this._socketId);
+        {
+          auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 2));
+          if (auto sharedWs = weakWs.lock())
+          {
+            sharedWs->Close(static_cast<IWebSocketResource::CloseCode>(jsArgAsInt(args, 0)), jsArgAsString(args, 1));
+          }
+        }
+        else if (args.size() == 1) // WebSocketModule.close(this._socketId);
+        {
+          auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 0));
+          if (auto sharedWs = weakWs.lock())
+          {
+            sharedWs->Close(IWebSocketResource::CloseCode::Normal, {});
+          }
+        }
+        else
+        {
+          auto errorObj = dynamic::object("id", -1)("message", "Incorrect number of parameters");
+          this->SendEvent("websocketFailed", std::move(errorObj));
+        }
+      }),
+    Method(
+      "send",
+      [this](dynamic args) // const string& message, int64_t id
+      {
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 1));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Send(jsArgAsString(args, 0));
+        }
+      }),
+    Method(
+      "sendBinary",
+      [this](dynamic args) // const string& base64String, int64_t id
+      {
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 1));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->SendBinary(jsArgAsString(args, 0));
+        }
+      }),
+    Method(
+      "ping",
+      [this](dynamic args) // int64_t id
+      {
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 0));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Ping();
+        }
+      })
+  };
 } // getMethods
 
 #pragma region private members
 
-void WebSocketModule::SendEvent(string &&eventName, dynamic &&args) {
+void WebSocketModule::SendEvent(string&& eventName, dynamic&& args)
+{
   auto instance = this->getInstance().lock();
   if (instance)
+  {
     instance->callJSFunction("RCTDeviceEventEmitter", "emit", dynamic::array(std::move(eventName), std::move(args)));
+  }
 }
 
-IWebSocketResource *WebSocketModule::GetOrCreateWebSocket(int64_t id, string &&url) {
-  IWebSocketResource *ptr = nullptr;
-  if (m_webSockets.find(id) == m_webSockets.end()) {
+std::weak_ptr<IWebSocketResource> WebSocketModule::GetOrCreateWebSocket(int64_t id, string&& url)
```

Comment:
>weak_ptr [](start = 5, length = 8)

Why this Api returns weak_ptr? Shouldn't it return shared_ptr? From shared_ptr, you can then convert to weak_ptr if needed

Reply:
Point taken. This PR has been submitted, but I may submit another one today and include that improvement.